src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java

```diff
@@ -190,44 +197,38 @@ public static void scan(Run build, PrintStream buildLog) {
      */
     private static List<FoundFailureCause> findCauses(final Collection<FailureCause> causes,
                                                       final Run build, final PrintStream buildLog) {
-        final List<FoundFailureCause> foundFailureCauseList =
-                Collections.synchronizedList(new LinkedList<FoundFailureCause>());
+        final List<FoundFailureCause> foundFailureCauseList = new ArrayList<FoundFailureCause>();
         long start = System.currentTimeMillis();
-        ThreadPoolExecutor executor = (ThreadPoolExecutor)Executors.
-                newFixedThreadPool(PluginImpl.getInstance().getNrOfScanThreads());
+
+        THREAD_POOL_EXECUTOR.setCorePoolSize(PluginImpl.getInstance().getNrOfScanThreads());
+        THREAD_POOL_EXECUTOR.setMaximumPoolSize(PluginImpl.getInstance().getNrOfScanThreads());
+
         buildLog.println("[BFA] Scanning build for known causes...");
-        for (final FailureCause cause : causes) {
-            executor.execute(new Runnable() {
+
+        Runnable runnable = new Runnable() {
                 @Override
                 public void run() {
-                    Thread.currentThread().setName("BFA-scanner-"
-                                                           + build.getFullDisplayName() + ": "
-                                                           + cause.getName());
-                    List<FoundIndication> foundIndications = findIndications(cause, build, buildLog);
-                    if (!foundIndications.isEmpty()) {
-                        FoundFailureCause foundFailureCause = new FoundFailureCause(cause, foundIndications);
-                        foundFailureCauseList.add(foundFailureCause);
-                    }
+                    Thread.currentThread().setName("BFA-scanner-" + build.getFullDisplayName());
+                    foundFailureCauseList.addAll(findIndications(causes, build, buildLog));
```

Comment:
The way I thought it worked was that you start one thread for all single lines, which is faster than going through the build log several times.

But this one thread is also used for handling everything in the old way, for Causes with at least one non-single-line indication. So in a setup with a lot of non-single-lines, what earlier was done in parallel in a lot of threads, is now done in a serial way, with one thread. 

This doesn't seem to be tweakable in any way for a specific build, only for the server as a whole. If any given log is long and you have lots of MultiLines, it will take a long time, is this correct or am I misreading something?